experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,962 @@
+/// the ready for the circuit on reset (e.g. once). To be refactored later
+class EndMemlessModel : public OpExecutionModel<handshake::EndOp> {
+public:
+  using OpExecutionModel<handshake::EndOp>::OpExecutionModel;
+  EndMemlessModel(
+      handshake::EndOp endOp,
+      mlir::DenseMap<std::pair<OpOperand *, Operation *>, RW *> &rws)
+      : OpExecutionModel<handshake::EndOp>(endOp), op(endOp) {
+    ins = getState<ChannelConsumerRW>(op.getInputs().front(), op, rws);
+    auto temp =
+        cast<handshake::ChannelType>(endOp.getInputs().front().getType());
+    bitwidth = temp.getDataBitWidth();
+  }
+
+  void exec(bool isReset) override {
+    if (isReset)
+      ins->ready = true;
+  }
+
+private:
+  handshake::EndOp op;
+  unsigned bitwidth;
+
+  ChannelConsumerRW *ins;
+};
+
+// Mutual component for binary operations: AddI, AndI, CmpI, MulI, OrI, ShlI,
+// ShrsI, ShruI, SubI, XorI
+using BinaryCompFunc = std::function<llvm::Any(llvm::Any, llvm::Any)>;
+template <typename Op>
+class GenericBinaryOpModel : public OpExecutionModel<Op> {
+public:
+  using OpExecutionModel<Op>::OpExecutionModel;
+  GenericBinaryOpModel(
+      Op op, mlir::DenseMap<std::pair<OpOperand *, Operation *>, RW *> &rws,
+      const BinaryCompFunc &callback, unsigned latency)
+      : OpExecutionModel<Op>(op), op(op), callback(callback), latency(latency),
+        bitwidth(cast<handshake::ChannelType>(op.getResult().getType())
+                     .getDataBitWidth()),
+        lhs(OpExecutionModel<Op>::template getState<ChannelConsumerRW>(
+            op.getLhs(), op, rws)),
+        rhs(OpExecutionModel<Op>::template getState<ChannelConsumerRW>(
+            op.getRhs(), op, rws)),
+        result(OpExecutionModel<Op>::template getState<ChannelProducerRW>(
+            op.getResult(), op, rws)),
+        binJoin(2) {}
+
+  void exec(bool isReset) override {
+    std::vector<bool> insValid{lhs->valid, rhs->valid};
+    std::vector<std::reference_wrapper<bool>> insReady{lhs->ready, rhs->ready};
+    // Set hadClock if the new iteration of the outer loop has started
+    if (!isReset) {
+      hadClock = true;
+    }
+
+    binJoin.exec(isReset, insValid, result->ready, insReady, result->valid);
+    // Store data internally if there were no data && lhs and rhs are valid
+    // (that is may transfer their data)
+    if (!hasData && result->valid && hadClock) {
+      hasData = true;
+      resData = any_cast<APInt>(callback(lhs->data, rhs->data));
+      result->valid = false;
+      counter = 0;
+    }
+    // Increase latency counter here
+    if (hasData && hadClock) {
+      ++counter;
+      hadClock = false;
+    }
+    // Now we are ready to transfer data
+    if (counter > latency) {
+      // Transfer is the consumer is ready to accept
+      if (result->ready)
+        result->valid = true;
+
+      hasData = false;
+      result->data = resData;
+    } else {
+      // latency
+      result->valid = false;
```

Comment:
This latency logic doesn't work unless the data arrives exactly once at the first simulation cycle. Data needs to come out exactly `latency` cycles after it arrives. The second time it arrives here it will come out `latency - 1` cycle after it arrives, until it ends up coming out immediately.